apache2/ag_mdb/ag_mdb_external.h

```diff
@@ -22,13 +22,30 @@ extern "C" {
  */
 
 /**
- ** Handler of AG-MDB, which contains a block of shared momory and a set of semaphores.
+ ** The structure stores the Read/Write locks information. 
+ ** In Linux, a set of semaphores is used to implement the Read/Write locks.
+ ** In Windows, The Windows Mutex is used to implement the Read/Write locks.
+ ** @param sem_id: The identifier of semaphore set in Linux.
+ ** @param read_lock_handle: The handler of the read locks in Windows.
+ ** @param write_lock_handle: The handler of the write lock in Windows.
+ */ 
+struct agmdb_lock{
+#ifndef _WIN32
+    int sem_id;
+#else
+    HANDLE read_lock_handle;
+    HANDLE write_lock_handle;
+#endif
+};
+
+/**
+ ** Handler of AG-MDB, which contains a block of shared momory and the lock structure.
  ** @param shm_base: The base address of the shared memory.
- ** @param sem_id: The identifier of semaphore set.
+ ** @param db_lock: The structure that stores the lock information.
  */ 
 struct agmdb_handler{
     const void* shm_base;
-    void* db_lock;
+    struct agmdb_lock db_lock;
```

Comment:
so, when we create the handler, let the system deal with the agmdb_lock memory allocation and deallocation..   We don't do it on our own due to hard to manage it?

Reply:
Yes. Because modsecurity use apr_pool to allocate memory, the deallocation operation is handled by apr library and modsecurity doesn't care about it. We can allocate memory for the db_lock, but we need to think through all the possible condition to deallocate it. If we leverage the apr_pool, we don't need to care about it.